python/lsst/daf/butler/queries/tree/_query_tree.py

```diff
@@ -0,0 +1,314 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = (
+    "QueryTree",
+    "make_identity_query_tree",
+    "DataCoordinateUploadKey",
+    "MaterializationKey",
+    "DatasetSearch",
+    "SerializedQueryTree",
+)
+
+import uuid
+from collections.abc import Mapping
+from typing import TypeAlias, final
+
+import pydantic
+
+from ...dimensions import DimensionGroup, DimensionUniverse
+from ...pydantic_utils import DeferredValidation
+from ._base import InvalidQueryError, QueryTreeBase
+from ._column_set import ColumnSet
+from ._predicate import Predicate
+
+DataCoordinateUploadKey: TypeAlias = uuid.UUID
+
+MaterializationKey: TypeAlias = uuid.UUID
+
+
+def make_identity_query_tree(universe: DimensionUniverse) -> QueryTree:
+    """Make an initial query tree with empty dimensions and a single logical
+    row.
+
+    This method should be used by `Butler._query` to construct the initial
+    query tree.  This tree is a useful initial state because it is the
+    identity for joins, in that joining any other query tree to the identity
+    yields that query tree.
+
+    Parameters
+    ----------
+    universe : `..DimensionUniverse`
+        Definitions for all dimensions.
+
+    Returns
+    -------
+    tree : `QueryTree`
+        A tree with empty dimensions.
+    """
+    return QueryTree(dimensions=universe.empty.as_group())
+
+
+@final
+class DatasetSearch(QueryTreeBase):
+    """Information about a dataset search joined into a query tree.
+
+    The dataset type name is the key of the dictionary (in `QueryTree`) where
+    this type is used as a value.
+    """
+
+    collections: tuple[str, ...]
+    """The collections to search.
+
+    Order matters if this dataset type is later referenced by a `FindFirst`
+    operation.  Collection wildcards are always resolved before being included
+    in a dataset search.
+    """
+
+    dimensions: DimensionGroup
+    """The dimensions of the dataset type.
+
+    This must match the dimensions of the dataset type as already defined in
+    the butler database, but this cannot generally be verified when a relation
+    tree is validated (since it requires a database query) and hence must be
+    checked later.
+    """
+
+    storage_class_name: str | None
+    """Name of the storage class to use when returning `DatasetRef` results.
+
+    May be `None` if the dataset is only used as a constraint or to return
+    columns that do not include a full dataset type.
+    """
```

Comment:
I struggled with this a lot myself.  I added it because:
 - if `Query.join_dataset_search` can take a `DatasetType` and the user wants dataset results, it'll be confusing if they passed in a storage class override initially and then we forgot it just because we didn't have a place to put it;
 - I wanted `Query.join_dataset_search` to be able to take `DatasetType` as well as `str`, both so `Query.datasets` could delegate to it and so I didn't have to document why it was unusual in that respect.

I think the real problem is that `DatasetType` has a storage class, and that's a problem in lots of other places, but of course it's tremendously useful in probably even more other places and having a `DatasetTypeWithoutStorageClass` public class is its own problem.  And that's all thoroughly water under the bridge.

In any case, yes, I think the server can ignore this (or just blindly copy it into Parquet files, so the client doesn't have to edit them); the important thing is that the client doesn't forget it, and I think it's slightly better (even if that means sending it to the server) than out on its own in `Query` (and it would be `Query`, not `SingleTypeDatasetResults`, to serve its current purpose).

Reply:
> if `Query.join_dataset_search` can take a `DatasetType` and the user wants dataset results, it'll be confusing if they passed in a storage class override initially and then we forgot it just because we didn't have a place to put it

Don't they already have to pass the dataset type again when they call `datasets()` to declare that they want refs back for this dataset type?  To me it makes sense that whatever you pass into `datasets` is the authoritative thing, since that's where you're declaring the output format.

>  not `SingleTypeDatasetResults`, to serve its current purpose).

It's possible I'm misunderstanding what is going on with this StorageClass (or maybe all the effects that a StorageClass can have?)  To me it looks like you could just pass in the StorageClass when you create the SingleTypeDatasetQueryResults, and call `ref.overrideStorageClass` if needed just before yielding from `__iter__`.  If this doesn't work I'd like to know because I'm using a similar pattern a few places in RemoteButler.

Reply:
Done on DM-43146: `DatasetSearch` no longer has `storage_class_name`.  Instead `join_dataset_search` raises if you give it a storage class override.